src/client/pythonEnvironments/base/locators/common/resourceBasedLocator.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { createDeferred, Deferred } from '../../../../common/utils/async';
+import { Disposables, IDisposable } from '../../../../common/utils/resourceLifecycle';
+import { PythonEnvInfo } from '../../info';
+import { IPythonEnvsIterator, Locator, PythonLocatorQuery } from '../../locator';
+
+export type Resource = IDisposable;
+
+/**
+ * A base locator class that manages the lifecycle of resources.
+ *
+ * The resources are not initialized until needed.
+ */
+export abstract class LazyResourceBasedLocator extends Locator implements IDisposable {
+    private readonly disposables = new Disposables();
+
+    // This will be set only once we have to create necessary resources
+    // and resolves once those resources are ready.
+    private resourcesReady?: Deferred<void>;
+
+    private watchersReady?: Deferred<void>;
+
+    public async dispose(): Promise<void> {
+        await this.disposables.dispose();
+    }
+
+    public async* iterEnvs(query?: PythonLocatorQuery): IPythonEnvsIterator {
+        await this.ensureResourcesReady();
+        await this.ensureWatchersReady();
```

Comment:
As mentioned in https://github.com/microsoft/vscode-python/pull/14713#discussion_r529929661, you do not need to await on watchers here.
```suggestion
        this.ensureWatchersReady().ignoreErrors();
```